File: core/SortList/SortList.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
    def merge_sort(node, length):
        if length <= 1:
            return node

        temp = node
        for i in range(0, length // 2 - 1):
            temp = temp.next

        next_node = temp.next
        temp.next = None
        left_node = merge_sort(node, length // 2)
        right_node = merge_sort(next_node, length - (length // 2))
        pre_head = ListNode()
        res = pre_head
        while left_node and right_node:
            if left_node.val <= right_node.val:
                pre_head.next = left_node
                left_node = left_node.next
            else:
                pre_head.next = right_node
                right_node = right_node.next
            pre_head = pre_head.next
        pre_head.next = left_node if left_node else right_node
        return res.next

    list_length = 0
    remain = head
    while head:
        list_length += 1
        head = head.next

    return merge_sort(remain, list_length)

File: core/SortList/SortList.java
package com.dbc;

public class SortList {
    public class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    private ListNode mergeSort(ListNode node, int length) {
        if (length <= 1) return node;
        ListNode remain = node;
        for (int i = 1; i < length / 2; i++) remain = remain.next;
        ListNode temp = remain.next;
        remain.next = null;
        ListNode leftNode = mergeSort(node, length / 2);
        ListNode rightNode = mergeSort(temp, length - length / 2);
        ListNode preHead = new ListNode();
        ListNode res = preHead;
        while (leftNode != null && rightNode != null) {
            if (leftNode.val <= rightNode.val) {
                preHead.next = leftNode;
                leftNode = leftNode.next;
            } else {
                preHead.next = rightNode;
                rightNode = rightNode.next;
            }
            preHead = preHead.next;
        }
        preHead.next = leftNode != null ? leftNode : rightNode;
        return res.next;
    }

    public ListNode sortList(ListNode head) {
        int length = 0;
        ListNode temp = head;
        while (temp != null) {
            temp = temp.next;
            length++;
        }
        return mergeSort(head, length);
    }
}
